20200401 New CKAN DataModel (#1850)
telefonicaid/fiware-cygnus: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_ckan_sink.md

[Top](#top)

#### <a name="section1.2.1"></a>Organizations naming conventions
* Data model by entity (`data_model=dm-by-entity`). An organization named as the notified `fiware-service` header value (or, in absence of such a header, the defaulted value for the FIWARE service) is created (if not existing yet).
Since based in [PostgreSQL only accepts] 
https://www.postgresql.org/docs/current/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS), it must be said only alphanumeric characters and the underscore (`_`) are accepted. The hyphen ('-') is also accepted. This leads to certain [encoding](#section2.3.3) is applied depending on the `enable_encoding` configuration parameter.
Nevertheless, different than PostgreSQL, [organization lengths](http://docs.ckan.org/en/latest/api/#ckan.logic.action.create.organization_create) may be up to 100 characters (minimum, 2 characters).

> * Data model by entity id (`data_model=dm-by-entity-id`). The organization name will take the value of the notified header `fiware-service` without encoding.

Reviewer: ```suggestion
* Data model by entity id (`data_model=dm-by-entity-id`). The organization name will take the value of the notified header `fiware-service`. Note that in this case, encoding is never applied (this is current marked as "debt" in the source code with a FIXME mark).
```

Reviewer 2: Fixed in [a0373a8](https://github.com/telefonicaid/fiware-cygnus/pull/1850/commits/a0373a8a5c2a17ab17f1703e953bb9b9598ddbed)

The following table summarizes the organization name composition:

| `dm-by-entity` | `dm-by-entity-id` |
|---|---|
| `<fiware-service>` | `<fiware-service>` | 

[Top](#top)

#### <a name="section1.2.2"></a>Packages/datasets naming conventions
* Data model by entity (`data_model=dm-by-entity`). A package/dataset named as the concatenation of the notified `fiware-service` and `fiware-servicePath` header values (or, in absence of such headers, the defaulted value for the FIWARE service and service path) is created (if not existing yet) in the above organization.
Since based in [PostgreSQL only accepts](https://www.postgresql.org/docs/current/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS), it must be said only alphanumeric characters and the underscore (`_`) are accepted. The hyphen ('-') is also accepted. This leads to  certain [encoding](#section2.3.3) is applied depending on the `enable_encoding` configuration parameter
Nevertheless, different than PostgreSQL, [dataset lengths](http://docs.ckan.org/en/latest/api/#ckan.logic.action.create.package_create) may be up to 100 characters (minimum, 2 characters).

> * Data model by entity id (`data_model=dm-by-entity-id`). A package/dataset name always take the entity ID. Such a name is already given in the NGSIEvent values, see the [Configuration](#section2.1) section for more details) within the the `NGSIEvent`.

Reviewer: ```suggestion
* Data model by entity id (`data_model=dm-by-entity-id`). A package/dataset name always take the entity ID. Such a name is already given in the NGSIEvent values, see the [Configuration](#section2.1) section for more details) within the the `NGSIEvent`. Note that in this case, encoding is never applied (this is current marked as "debt" in the source code with a FIXME mark).
```

Reviewer 2: Fixed in [4a57860](https://github.com/telefonicaid/fiware-cygnus/pull/1850/commits/4a57860810552aee2b19859af7f21d61baa8f3ba)

The following table summarizes the package name composition:

| FIWARE service path | `dm-by-entity` | `dm-by-entity-id` |
|---|---|---|
| `/` | `<fiware-service>_<fiware-servicePath>` | `<entityId>` | 

[Top](#top)

#### <a name="section1.2.3"></a>Resources naming conventions

> The resiurce name depends on the configured data model (see the [Configuration](#section2.1) section for more details):

Reviewer: ```suggestion
The resource name depends on the configured data model (see the [Configuration](#section2.1) section for more details):
```

Reviewer 2: Fixed in [db4c369](https://github.com/telefonicaid/fiware-cygnus/pull/1850/commits/db4c369038e851be614e0c5e15585f31d3a1303a)

* Data model by entity (`data_model=dm-by-entity`). A resource name always take the concatenation of the entity ID and type. Such a name is already given in the `notified_entities`/`grouped_entities` header values (depending on using or not the grouping rules, see the [Configuration](#section2.1) section for more details) within the `NGSIEvent`.

> * Data model by entity id (`data_model=dm-by-entity-id`). A resource name always take the entity ID. Such a name is already given in the NGSIEvent values, see the [Configuration](#section2.1) section for more details) within the the `NGSIEvent`. 

Reviewer: ```suggestion
* Data model by entity id (`data_model=dm-by-entity-id`). A resource name always take the entity ID. Such a name is already given in the NGSIEvent values, see the [Configuration](#section2.1) section for more details) within the the `NGSIEvent`. Note that in this case, encoding is never applied (this is current marked as "debt" in the source code with a FIXME mark).
```

Reviewer 2: Fixed in [176c876](https://github.com/telefonicaid/fiware-cygnus/pull/1850/commits/176c8767ac28365ce902adf995661353e21ac7f7)

It must be noticed a CKAN Datastore (and a viewer) is also created and associated to the resource above. This datastore, which in the end is a PostgreSQL table, will hold the persisted data.

Since based in [PostgreSQL](https://www.postgresql.org/docs/current/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS), it must be said only alphanumeric characters and the underscore (`_`) are accepted. The hyphen ('-') is also accepted. This leads to  certain [encoding](#section2.3.3) is applied depending on the `enable_encoding` configuration parameter.

Despite there is no real limit on the resource names, Cygnus will keep limiting their lengths up to 100 characters (minimum, 2 characters), accordingly to what's done with organization and package names.

The following table summarizes the resource name composition:

| FIWARE service path | `dm-by-entity` | `dm-by-entity-id` |
|---|---|---|

> | `/` | `<entityType>` | `<entityId>` |

Reviewer: According to " Data model by entity (`data_model=dm-by-entity`). A resource name always take the concatenation of the entity ID and type", is the second one (`<entityType>`) correct? Shouldn't include the entity id concatened in some way?

Reviewer 2: This is interesting. On the [actuall implementation of the resource building name](https://github.com/telefonicaid/fiware-cygnus/blob/master/cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSICKANSink.java#L445) It's used the entity, not the concatenation of the entity ID and type as mentioned here. 

I'll update this doc according to that, do you agree?

| enable\_encoding | no | false | <i>true</i> or <i>false</i>, <i>true</i> applies the new encoding, <i>false</i> applies the old encoding. ||
| enable\_grouping | no | false | <i>true</i> or <i>false</i>. Check this [link](./ngsi_grouping_interceptor.md) for more details. ||
| enable\_name\_mappings | no | false | <i>true</i> or <i>false</i>. Check this [link](./ngsi_name_mappings_interceptor.md) for more details. ||

> | data\_model | no | dm-by-entity | <i>dm-by-entity-id</i>, <i>dm-by-entity</i> ||

Reviewer: Previously to this PR, dm-by-entity was the only mode supported, wasn't it? Or, in other words, this PR has added the possibility of using dm-by-entity-id.

Is my understanding correct?

Author: Yes , its correct

Reviewer: Thanks for the clarification!

NTC

> | data\_model | no | dm-by-entity | dm-by-entity-id | <i>dm-by-entity</i>, even if not configured. ||

Reviewer 2: | data\_model | no | dm-by-entity | <i>dm-by-entity-id</i>, <i>dm-by-entity</i> ||
```suggestion
| data\_model | no | dm-by-entity | dm-by-entity-id | <i>dm-by-entity</i>, even if not configured. ||
```

Author: Solved in aad6fe1d70613f27b5bae4ab36969c8fa512c61c
